File: moxy-compiler/src/main/java/com/omegar/mvp/compiler/pipeline/ElementsAnnotatedGenerator.java
package com.omegar.mvp.compiler.pipeline;

import com.omegar.mvp.compiler.MvpCompiler;
import com.omegar.mvp.compiler.entity.AnnotationInfo;

import java.util.Set;

import javax.annotation.processing.Messager;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.tools.Diagnostic;

/**
 * Created by Anton Knyazev on 03.12.2020.
 */
public class ElementsAnnotatedGenerator<T extends Element> extends Publisher<T> {

    private final RoundEnvironment mRoundEnv;
    private final Messager mMessager;
    private final AnnotationInfo<T> mAnnotationInfo;

    public ElementsAnnotatedGenerator(RoundEnvironment roundEnv, Messager messager, AnnotationInfo<T> annotationInfo) {
        mRoundEnv = roundEnv;
        mMessager = messager;
        mAnnotationInfo = annotationInfo;
    }

    @Override
    public void publish(PipelineContext<T> context) {
        ElementKind elementKind = mAnnotationInfo.getElementKind();
        Set<? extends Element> allElements = mRoundEnv.getElementsAnnotatedWith(mAnnotationInfo.getAnnotationTypeElement());

        for (Element element : allElements) {
            if (element.getKind() != elementKind) {
                mMessager.printMessage(Diagnostic.Kind.ERROR,
                        element + " must be " + elementKind.name()
                                + ", or not mark it as @" + mAnnotationInfo.getAnnotationTypeElement().getSimpleName());
            } else {
                //noinspection unchecked
                context.next((T) element);
            }
        }
        finish(context);
    }

}

File: moxy/compiler/src/main/java/com/omegar/mvp/compiler/pipeline/ElementsAnnotatedGenerator.kt
package com.omegar.mvp.compiler.pipeline

import com.omegar.mvp.compiler.entity.AnnotationInfo
import javax.annotation.processing.Messager
import javax.annotation.processing.RoundEnvironment
import javax.lang.model.element.Element
import javax.tools.Diagnostic

/**
 * Created by Anton Knyazev on 03.12.2020.
 */
class ElementsAnnotatedGenerator<T : Element>(
    private val mRoundEnv: RoundEnvironment,
    private val mMessager: Messager,
    private val annotationInfo: AnnotationInfo<T>
) : Publisher<T>() {

    override fun publish(context: PipelineContext<T>?) {
        val elementKind = annotationInfo.elementKind
        val allElements = mRoundEnv.getElementsAnnotatedWith(annotationInfo.typeElement)
        for (element in allElements) {
            if (element.kind != elementKind) {
                mMessager.printMessage(
                    Diagnostic.Kind.ERROR,
                    "$element must be ${elementKind.name}, or not mark it as @$annotationInfo"
                )
            } else {
                @Suppress("UNCHECKED_CAST")
                context?.next(element as T)
            }
        }
        finish(context)
    }

}